### PR TITLE
queue blocks token loader AGPUSH-1521

### PR DESCRIPTION
**Pull request by lfryc**
Use address-full-policy=FAIL instead of PAGE,
means transactions will fail and they will be retried after configured delay,
instead of paging to disk and eventually failing on running out of disk space.

And upgrade to WildFly 10.

https://issues.jboss.org/browse/AGPUSH-1521

### CODE REVIEW

File: configuration/jms-setup-wildfly.cli [hunk 1]
```diff
@@ -1,9 +1,16 @@
 connect
+
+#/subsystem=logging/console-handler=AEROGEAR:add(formatter="%d{HH:mm:ss,SSS} %-5p [%c] (%t) %s%E%n",autoflush=true)
+#/subsystem=logging/logger=org.jboss.aerogear.unifiedpush:add(level=FINEST,use-parent-handlers=false,handlers=["AEROGEAR"])
+
```
Comment (lfryc) [inline configuration/jms-setup-wildfly.cli]: needs removal

File: servers/ups-wildfly/src/main/webapp/WEB-INF/hornetq-jms.xml [hunk 2]
```diff
@@ -82,5 +82,5 @@
                 <entry name="/queue/AllBatchesLoadedQueue"/>
             </jms-queue>
         </jms-destinations>
-    </hornetq-server>
+    </server>
 </messaging-deployment>
```
Comment (lfryc) [inline servers/ups-wildfly/src/main/webapp/WEB-INF/hornetq-jms.xml]: this is specific to AMQ Artemis - we should set queues with CLI script (same as we do with address-settings) and thus we will have two cli scripts (one for HornetQ, one for AMQ)... at the end, we can remove hornetq-jms.xml

**Comment by lfryc** [comment]
I've tested with WildFly 10.0.0.CR5 and it Android push notification was delivered to real device

**Comment by lfryc** [comment]
There are test failures, I will look into them shortly, but in the meantime, this can be tested manually

**Comment by matzew** [comment]
:eyes: 

btw. why WF-10-CR5? Final was released two month ago :)

**Comment by lfryc** [comment]
@matzew yea, just my commits are 2 months around :-)

**Comment by matzew** [comment]
@lfryc tests are failing, locally and on Travis

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/NotificationRouter.java [hunk 3]
```diff
@@ -55,7 +56,7 @@
     private final AeroGearLogger logger = AeroGearLogger.getInstance(NotificationRouter.class);
 
     @Inject
-    private GenericVariantService genericVariantService;
+    private Instance<GenericVariantService> genericVariantService;
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/NotificationRouter.java]: why that ?

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java [hunk 4]
```diff
@@ -169,4 +188,42 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
...
+     *
+     * @return returns true if token batch was queued; return false is queue was still full even after retrying
+     */
+    private boolean tryToDispatchTokens(MessageHolderWithTokens msg) throws InterruptedException {
+        for (int i = 0 ; i <= MAX_TIME_TO_BLOCK_WHEN_QUEUE_IS_FULL; i += TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL) {
+            if (dispatchTokensOrFailWhenQueueIsFull(msg)) {
+                return true;
+            } else {
+                logger.fine(String.format("Queue is full, blocking loading of batch token #%s for %s variant (%s) for %s ms", msg.getSerialId(), msg.getVariant().getType().getTypeName(), msg.getVariant().getVariantID(), TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL));
+                Thread.sleep(TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL);
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: is this really a good idea?

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java [hunk 5]
```diff
@@ -169,4 +188,42 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
...
+     *
+     * @return returns true if tokens were successfully queued; returns false if queue was full
+     */
+    private boolean dispatchTokensOrFailWhenQueueIsFull(MessageHolderWithTokens msg) {
+        try {
+            dispatchTokensEvent.fire(msg);
+            return true;
+        } catch (MessageDeliveryException e) {
+            Throwable cause = e.getCause();
+            if (cause.getMessage() != null && cause.getMessage().contains("is full")) {
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: perhaps a bit more type safty ?

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/holder/MessageHolderWithVariants.java [hunk 6]
```diff
@@ -34,6 +34,8 @@
 
     private static final long serialVersionUID = -7955411139315335655L;
 
+    public static final int INITIAL_SERIAL_ID = 0;
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/holder/MessageHolderWithVariants.java]: what is this?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    @Inject
+    private JmsClient jmsClient;
+
+    @Produces
+    public ClientInstallationService getClientInstallationService() {
+        return new ClientInstallationService() {
+
+            @Override
+            public void updateInstallation(Installation toUpdate, Installation postedInstallation) {
```
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java]: how about jmock? or something like that ?

File: servers/ups-wildfly/src/main/webapp/WEB-INF/hornetq-jms.xml (hunk 2 above)
Comment (matzew) [inline servers/ups-wildfly/src/main/webapp/WEB-INF/hornetq-jms.xml]: this also fixes: AGPUSH-1587

File: configuration/jms-setup-wildfly.cli [hunk 8]
```diff
@@ -1,9 +1,16 @@
...
 batch
 
-/subsystem=messaging/hornetq-server=default/address-setting=jms.queue.TriggerMetricCollectionQueue:add(redelivery-delay=1000, max-delivery-attempts=300)
-/subsystem=messaging/hornetq-server=default/address-setting=jms.queue.MetricsQueue:add(max-delivery-attempts=-1)
-/subsystem=messaging/hornetq-server=default/address-setting=jms.queue.BatchLoadedQueue:add(max-delivery-attempts=-1)
-/subsystem=messaging/hornetq-server=default/address-setting=jms.queue.AllBatchesLoadedQueue:add(max-delivery-attempts=-1)
+/subsystem=messaging-activemq/server=default/address-setting=jms.queue.GCMPushMessageQueue:add(redelivery-delay=1500, redelivery-multiplier=1.5, max-redelivery-delay=5000, max-delivery-attempts=300)
+/subsystem=messaging-activemq/server=default/address-setting=jms.queue.GCMTokenBatchQueue:add(address-full-policy=FAIL, max-size-bytes=40000)
+/subsystem=messaging-activemq/server=default/address-setting=jms.queue.TriggerMetricCollectionQueue:add(redelivery-delay=1000, max-delivery-attempts=300)
+
```
Comment (matzew) [inline configuration/jms-setup-wildfly.cli]: Is it possible to have WF log something? when we add a Datasource, the servr console tells us it was added, besides just the CLI saying: Yup, got it

File: configuration/jms-setup-wildfly.cli [hunk 9]
```diff
@@ -1,9 +1,16 @@
 connect
+
+/subsystem=logging/console-handler=AEROGEAR:add(formatter="%d{HH:mm:ss,SSS} %-5p [%c] (%t) %s%E%n",autoflush=true)
+/subsystem=logging/logger=org.jboss.aerogear.unifiedpush:add(level=FINEST,use-parent-handlers=false,handlers=["AEROGEAR"])
```
Comment (matzew) [inline configuration/jms-setup-wildfly.cli]: this is only for better tests, right ?

Comment (matzew) [inline configuration/jms-setup-wildfly.cli]: before merging, let's please comment this out

Comment (lfryc) [inline configuration/jms-setup-wildfly.cli]: done

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/NotificationRouter.java (hunk 3 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/NotificationRouter.java]: it was causing a test error - the mock genericVariantService was stored as `@RequestScoped` but the bean was seeing the cached version - I'm not sure why, but using `Instance#get()` refreshed the instance

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 4 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: Not sure it is good idea, I was testing three options:

1) using `address-full-policy=BLOCK` blocks the thread until the queue has free space, which is resource wasting

2) using `address-full-policy=FAIL` will immediatelly fail MDB transaction, so the message will be redelivered (according to CLI settings, it will try it in 1 second, with exponential backoff 1.5x with each next failure)

I don't believe we need to fail immediatelly, since we may instead block for reasonable time and if the queue still hasn't space available, we will fail MDB transaction anyway. So last option I went with at the end is a hybrid:

3) block for one second and if queue is still full, we will fail (so we don't block too long, but we give it some time free space)

May be premature optimization though.

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 5 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: yea, we would have to go implementation-specific, Hornetq/ActiveMQ use different classes

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/holder/MessageHolderWithVariants.java (hunk 6 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/holder/MessageHolderWithVariants.java]: I will add comment, that's the serialId for first token batch. We use it in tests, so I wanted to avoid that be a magic constant

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java (hunk 7 above)
Comment (lfryc) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java]: yea, we should implement just the method that's needed.

File: configuration/jms-setup-wildfly.cli (hunk 8 above)
Comment (lfryc) [inline configuration/jms-setup-wildfly.cli]: I believe we should move queue/topic creation to CLI script - that way it should print the queue/topic creation info.

Comment (lfryc) [inline configuration/jms-setup-wildfly.cli]: yea, when we deploy through CLI:

````
jms-queue add --queue-address=AdmPushMessageQueue --entries=queue/AdmPushMessageQueue
````

it prints:

````
14:28:47,224 INFO  [org.apache.activemq.artemis.core.server] (ServerService Thread Pool -- 12) AMQ221003: trying to deploy queue jms.queue.AdmPushMessageQueue
````

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/NotificationRouter.java (hunk 3 above)
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/NotificationRouter.java]: ok - that's fine, than

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 4 above)
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: @thradec mind taking a look here ?

Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: > I don't believe we need to fail immediatelly,

that's right 

not exactly sure, what's really the best

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 5 above)
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: ok... can you add a comment ? with that info?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java (hunk 7 above)
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java]: @lfryc wanna do that in a different PR, or JIRA?

File: configuration/jms-setup-wildfly.cli (hunk 8 above)
Comment (matzew) [inline configuration/jms-setup-wildfly.cli]: @lfryc can you do that please?

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 4 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: ok, we agreed offline to remove this to improve readability

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 5 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: yep, I will make it more clear what we are doing there

**Comment by lfryc** [comment]
@matzew ok, all comments were incorporated :recycle:

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java [hunk 10]
```diff
@@ -190,40 +190,32 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
...
-     *
-     * @return returns true if token batch was queued; return false is queue was still full even after retrying
-     */
-    private boolean tryToDispatchTokens(MessageHolderWithTokens msg) throws InterruptedException {
-        for (int i = 0 ; i <= MAX_TIME_TO_BLOCK_WHEN_QUEUE_IS_FULL; i += TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL) {
-            if (dispatchTokensOrFailWhenQueueIsFull(msg)) {
-                return true;
-            } else {
-                logger.fine(String.format("Queue is full, blocking loading of batch token #%s for %s variant (%s) for %s ms", msg.getSerialId(), msg.getVariant().getType().getTypeName(), msg.getVariant().getVariantID(), TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL));
-                Thread.sleep(TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL);
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: please remove the `TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL` and `MAX_TIME_TO_BLOCK_WHEN_QUEUE_IS_FULL ` form the class

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java [hunk 11]
```diff
@@ -190,40 +190,32 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
...
+
+    /**
+     * Since different messaging implementations use different exceptions when queue is full,
+     * we have to detect that queue is full by analyzing properties of the thrown exception.
+     *
+     * @param e throwable thrown when JMS message delivery fails
+     * @return true if exceptions represents state when queue is full; false otherwise
+     */
+    private boolean isQueueFullException(Throwable e) {
+        return e != null && e.getMessage().contains("is full");
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: what's the exact message? what about German? :) Perhaps better to check for class name(s)?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java [hunk 12]
```diff
@@ -42,62 +43,19 @@
...
-            }
-
-            @Override
-            public void removeInstallationForVariantByDeviceToken(String variantID, String deviceToken) {
-                // TODO Auto-generated method stub
-
-            }
+        ClientInstallationService mock = mock(ClientInstallationService.class);
 
+        when(mock.findAllDeviceTokenForVariantIDByCriteria(Mockito.anyString(), Mockito.anyList(), Mockito.anyList(), Mockito.anyList(), Mockito.anyInt(), Mockito.anyString())).thenAnswer(new Answer<QueryBuilder<String>>() {
```
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/MocksForTokenLoaderTransactionFailForGCM.java]: :+1:

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 10 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: nice catch @matzew !

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java (hunk 11 above)
Comment (lfryc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java]: :+1:

**Comment by matzew** [comment]
@lfryc :tada: thanks !

I did remove the unsed bits myself, can you follow up on: https://github.com/aerogear/aerogear-unifiedpush-server/pull/694#discussion_r56679147

in a different/new PR?
